!["RESIM"](flo_rfm.png)

!["RESIM"](flo_veri_hikayesi.png)

## Veri Setini Okuma ve Veri Seti Hakkında Bilgi Edinme

- Gereki kütüphanelerin Kurulması

In [1]:
import datetime as dt
import pandas as pd
pd.set_option('display.max_columns', None)#Ekran ayarı
pd.set_option('display.max_rows', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)#ondalıklı sayılarda,noktadan sonra 3 basamak yaz.



In [2]:
df_ = pd.read_csv("flo_data_20k.csv")

In [3]:
df = df_.copy()

In [4]:
def data_sum(df):
    print("################################ VERİ ÖN HAZIRLIK SÜRECİ ####################################","\n",
      "1-İlk 10 Gözlem:","\n\n",df.head(10),"\n\n",
      "2-Değişken İsimleri:","\n\n",df.columns,"\n\n",
      "3-Betimsel İstatistik:","\n\n",df.describe().T,"\n\n",
      "4:Boş Gözlem Sayısı:","\n\n",df.isnull().sum(),"\n\n",
      "5-Değişkenlerin Türü:","\n\n",df.dtypes,"\n\n")
   

In [5]:
data_sum(df)

################################ VERİ ÖN HAZIRLIK SÜRECİ #################################### 
 1-İlk 10 Gözlem: 

                               master_id order_channel last_order_channel  \
0  cc294636-19f0-11eb-8d74-000d3a38a36f   Android App            Offline   
1  f431bd5a-ab7b-11e9-a2fc-000d3a38a36f   Android App             Mobile   
2  69b69676-1a40-11ea-941b-000d3a38a36f   Android App        Android App   
3  1854e56c-491f-11eb-806e-000d3a38a36f   Android App        Android App   
4  d6ea1074-f1f5-11e9-9346-000d3a38a36f       Desktop            Desktop   
5  e585280e-aae1-11e9-a2fc-000d3a38a36f       Desktop            Offline   
6  c445e4ee-6242-11ea-9d1a-000d3a38a36f   Android App        Android App   
7  3f1b4dc8-8a7d-11ea-8ec0-000d3a38a36f        Mobile            Offline   
8  cfbda69e-5b4f-11ea-aca7-000d3a38a36f   Android App        Android App   
9  1143f032-440d-11ea-8b43-000d3a38a36f        Mobile             Mobile   

  first_order_date last_order_date last_order_d

- Omnichannel müşterilerin hem online'dan hemde offline platformlardan alışveriş yaptığını ifade etmektedir. Her bir müşterinin toplam alışveriş sayısı ve harcaması için yeni değişkenler oluşturuyoruz.

In [6]:
df["total_shopping"]= df["order_num_total_ever_online"] + df["order_num_total_ever_offline"]
df["total_shopping"].head()

0    5.000
1   21.000
2    5.000
3    2.000
4    2.000
Name: total_shopping, dtype: float64

In [7]:
df["total_value"]= df["customer_value_total_ever_offline"] + df["customer_value_total_ever_online"]
df["total_value"].head()

0    939.370
1   2013.550
2    585.320
3    121.970
4    209.980
Name: total_value, dtype: float64

In [8]:
df.head()

master_id order_channel last_order_channel  \
0  cc294636-19f0-11eb-8d74-000d3a38a36f   Android App            Offline   
1  f431bd5a-ab7b-11e9-a2fc-000d3a38a36f   Android App             Mobile   
2  69b69676-1a40-11ea-941b-000d3a38a36f   Android App        Android App   
3  1854e56c-491f-11eb-806e-000d3a38a36f   Android App        Android App   
4  d6ea1074-f1f5-11e9-9346-000d3a38a36f       Desktop            Desktop   

  first_order_date last_order_date last_order_date_online  \
0       2020-10-30      2021-02-26             2021-02-21   
1       2017-02-08      2021-02-16             2021-02-16   
2       2019-11-27      2020-11-27             2020-11-27   
3       2021-01-06      2021-01-17             2021-01-17   
4       2019-08-03      2021-03-07             2021-03-07   

  last_order_date_offline  order_num_total_ever_online  \
0              2021-02-26                        4.000   
1              2020-01-10                       19.000   
2              2019-12-01                        3.000   
3              2021-01-06                        1.000   
4              2019-08-03                        1.000   

   order_num_total_ever_offline  customer_value_total_ever_offline  \
0                         1.000                            139.990   
1                         2.000                            159.970   
2                         2.000                            189.970   
3                         1.000                             39.990   
4                         1.000                             49.990   

   customer_value_total_ever_online       interested_in_categories_12  \
0                           799.380                           [KADIN]   
1                          1853.580  [ERKEK, COCUK, KADIN, AKTIFSPOR]   
2                           395.350                    [ERKEK, KADIN]   
3                            81.980               [AKTIFCOCUK, COCUK]   
4                           159.990                       [AKTIFSPOR]   

   total_shopping  total_value  
0           5.000      939.370  
1          21.000     2013.550  
2           5.000      585.320  
3           2.000      121.970  
4           2.000      209.980

- Değişken tiplerini incelediğimizde tarih ifade eden değişkenin tipinin object olduğunu görüyoruz.Veri tipini date'e çeviriyoruz

In [9]:
df.dtypes

master_id                             object
order_channel                         object
last_order_channel                    object
first_order_date                      object
last_order_date                       object
last_order_date_online                object
last_order_date_offline               object
order_num_total_ever_online          float64
order_num_total_ever_offline         float64
customer_value_total_ever_offline    float64
customer_value_total_ever_online     float64
interested_in_categories_12           object
total_shopping                       float64
total_value                          float64
dtype: object

In [10]:
date_columns = df.columns[df.columns.str.contains("date")]
df[date_columns] = df[date_columns].apply(pd.to_datetime)

In [11]:
df.dtypes

master_id                                    object
order_channel                                object
last_order_channel                           object
first_order_date                     datetime64[ns]
last_order_date                      datetime64[ns]
last_order_date_online               datetime64[ns]
last_order_date_offline              datetime64[ns]
order_num_total_ever_online                 float64
order_num_total_ever_offline                float64
customer_value_total_ever_offline           float64
customer_value_total_ever_online            float64
interested_in_categories_12                  object
total_shopping                              float64
total_value                                 float64
dtype: object

- Alışveriş kanallarındaki müşteri sayısının, toplam alınan ürün sayısının ve toplam harcamaların dağılımına bakıyoruz.

In [12]:
df.groupby("order_channel").agg({"master_id":"count",
                                 "total_shopping":"sum",
                                 "total_value":"sum"})


master_id  total_shopping  total_value
order_channel                                        
Android App         9495       52269.000  7819062.760
Desktop             2735       10920.000  1610321.460
Ios App             2833       15351.000  2525999.930
Mobile              4882       21679.000  3028183.160

- En fazla kazancı getiren ilk 10 müşteriyi sıralayalım.

In [13]:
max_kazanc = df.groupby("master_id").agg({"total_value" :"sum"}).sort_values("total_value",ascending = False)
max_kazanc.head(10)

total_value
master_id                                        
5d1c466a-9cfd-11e9-9897-000d3a38a36f    45905.100
d5ef8058-a5c6-11e9-a2fc-000d3a38a36f    36818.290
73fd19aa-9e37-11e9-9897-000d3a38a36f    33918.100
7137a5c0-7aad-11ea-8f20-000d3a38a36f    31227.410
47a642fe-975b-11eb-8c2a-000d3a38a36f    20706.340
a4d534a2-5b1b-11eb-8dbd-000d3a38a36f    18443.570
d696c654-2633-11ea-8e1c-000d3a38a36f    16918.570
fef57ffa-aae6-11e9-a2fc-000d3a38a36f    12726.100
cba59206-9dd1-11e9-9897-000d3a38a36f    12282.240
fc0ce7a4-9d87-11e9-9897-000d3a38a36f    12103.150

- En fazla siparişi veren ilk 10 müşteriyi sıralayalım.

In [14]:
max_siparis = df.groupby("master_id").agg({"total_shopping" :"sum"}).sort_values("total_shopping",ascending = False)
max_siparis.head(10)

total_shopping
master_id                                           
5d1c466a-9cfd-11e9-9897-000d3a38a36f         202.000
cba59206-9dd1-11e9-9897-000d3a38a36f         131.000
a57f4302-b1a8-11e9-89fa-000d3a38a36f         111.000
fdbe8304-a7ab-11e9-a2fc-000d3a38a36f          88.000
329968c6-a0e2-11e9-a2fc-000d3a38a36f          83.000
73fd19aa-9e37-11e9-9897-000d3a38a36f          82.000
44d032ee-a0d4-11e9-a2fc-000d3a38a36f          77.000
b27e241a-a901-11e9-a2fc-000d3a38a36f          75.000
d696c654-2633-11ea-8e1c-000d3a38a36f          70.000
a4d534a2-5b1b-11eb-8dbd-000d3a38a36f          70.000

- Veri ön hazırlık sürecini fonksiyonlaştıralım.

# **2:** RFM Metriklerinin Hesaplanması

- **Recency =** Müşterinin en son alışverişi tarihini gösterir.
- **Frequency =** Müşterinin alışveriş sıklığını gösterir.
- **Monetary =** Müşterinin bize bıraktığı parasal değer.

- Müşteri özelinde Recency, Frequency ve Monetary metriklerini hesaplayoruz.

In [15]:
df.sort_values(by="last_order_date",ascending = False).head(2)

master_id order_channel last_order_channel  \
12438  625f40a2-5bd2-11ea-98b0-000d3a38a36f   Android App        Android App   
19111  3991df04-d707-11e9-93bc-000d3a38a36f       Ios App            Ios App   

      first_order_date last_order_date last_order_date_online  \
12438       2019-12-27      2021-05-30             2021-05-30   
19111       2019-09-12      2021-05-30             2021-05-30   

      last_order_date_offline  order_num_total_ever_online  \
12438              2019-12-27                       15.000   
19111              2021-04-26                        1.000   

       order_num_total_ever_offline  customer_value_total_ever_offline  \
12438                         1.000                            229.980   
19111                         2.000                            319.970   

       customer_value_total_ever_online     interested_in_categories_12  \
12438                          9723.330  [AKTIFCOCUK, KADIN, AKTIFSPOR]   
19111                           319.990                     [AKTIFSPOR]   

       total_shopping  total_value  
12438          16.000     9953.310  
19111           3.000      639.960

In [16]:
today_date = df["last_order_date"].max() + dt.timedelta(days=2)
today_date

Timestamp('2021-06-01 00:00:00')

In [17]:
rfm = df.groupby('master_id').agg({'last_order_date': lambda last_order_date: (today_date - last_order_date.max()).days,
                             'total_shopping': lambda total_shopping: total_shopping,
                             'total_value': lambda total_value: total_value})


In [18]:
rfm.columns = ['recency', 'frequency', 'monetary']

In [19]:
rfm.head()

recency  frequency  monetary
master_id                                                         
00016786-2f5a-11ea-bb80-000d3a38a36f       10      5.000   776.070
00034aaa-a838-11e9-a2fc-000d3a38a36f      298      3.000   269.470
000be838-85df-11ea-a90b-000d3a38a36f      213      4.000   722.690
000c1fe2-a8b7-11ea-8479-000d3a38a36f       27      7.000   874.160
000f5e3e-9dde-11ea-80cd-000d3a38a36f       20      7.000  1620.330

### RFM Skorunun Hesaplanması

 Recency, Frequency ve Monetary metriklerini qcut yardımı ile 1-5 arasında skorlara çevirerek bir standartlaştırma işlemi yapmış bulunuyoruz.


In [20]:
rfm["recency_score"] = pd.qcut(rfm['recency'], 5, labels=[5, 4, 3, 2, 1])

rfm["frequency_score"] = pd.qcut(rfm['frequency'].rank(method="first"), 5, labels=[1, 2, 3, 4, 5])

rfm["monetary_score"] = pd.qcut(rfm['monetary'], 5, labels=[1, 2, 3, 4, 5])


**Adım 3:** recency_score ve frequency_score’u tek bir değişken olarak ifade ediniz ve RF_SCORE olarak kaydediniz.

In [21]:
rfm["RF_SCORE"] = (rfm['recency_score'].astype(str) +
                    rfm['frequency_score'].astype(str))
rfm["RF_SCORE"].head()

master_id
00016786-2f5a-11ea-bb80-000d3a38a36f    54
00034aaa-a838-11e9-a2fc-000d3a38a36f    12
000be838-85df-11ea-a90b-000d3a38a36f    23
000c1fe2-a8b7-11ea-8479-000d3a38a36f    54
000f5e3e-9dde-11ea-80cd-000d3a38a36f    54
Name: RF_SCORE, dtype: object

###  RF Skorunun Segment Olarak Tanımlanması

 Oluşturulan RF skorları için segment tanımlamaları yapalım.

In [22]:
seg_map = {
    r'[1-2][1-2]': 'hibernating',
    r'[1-2][3-4]': 'at_Risk',
    r'[1-2]5': 'cant_loose',
    r'3[1-2]': 'about_to_sleep',
    r'33': 'need_attention',
    r'[3-4][4-5]': 'loyal_customers',
    r'41': 'promising',
    r'51': 'new_customers',
    r'[4-5][2-3]': 'potential_loyalists',
    r'5[4-5]': 'champions'
}

Yukarıda tanımlamış olduğumuz seg_map yardımı ile skorları segmentlere çeviriyoruz.

In [23]:
rfm['segment'] = rfm['RF_SCORE'].replace(seg_map, regex=True)

rfm.head()

recency  frequency  monetary  \
master_id                                                            
00016786-2f5a-11ea-bb80-000d3a38a36f       10      5.000   776.070   
00034aaa-a838-11e9-a2fc-000d3a38a36f      298      3.000   269.470   
000be838-85df-11ea-a90b-000d3a38a36f      213      4.000   722.690   
000c1fe2-a8b7-11ea-8479-000d3a38a36f       27      7.000   874.160   
000f5e3e-9dde-11ea-80cd-000d3a38a36f       20      7.000  1620.330   

                                     recency_score frequency_score  \
master_id                                                            
00016786-2f5a-11ea-bb80-000d3a38a36f             5               4   
00034aaa-a838-11e9-a2fc-000d3a38a36f             1               2   
000be838-85df-11ea-a90b-000d3a38a36f             2               3   
000c1fe2-a8b7-11ea-8479-000d3a38a36f             5               4   
000f5e3e-9dde-11ea-80cd-000d3a38a36f             5               4   

                                     monetary_score RF_SCORE      segment  
master_id                                                                  
00016786-2f5a-11ea-bb80-000d3a38a36f              4       54    champions  
00034aaa-a838-11e9-a2fc-000d3a38a36f              1       12  hibernating  
000be838-85df-11ea-a90b-000d3a38a36f              4       23      at_Risk  
000c1fe2-a8b7-11ea-8479-000d3a38a36f              4       54    champions  
000f5e3e-9dde-11ea-80cd-000d3a38a36f              5       54    champions

In [24]:
rfm.groupby("segment").agg({"recency":["mean","count"],
                            "frequency":["mean","count"],
                            "monetary":["mean","count"]}) 


recency       frequency       monetary      
                       mean count      mean count     mean count
segment                                                         
about_to_sleep      114.032  1643     2.407  1643  361.649  1643
at_Risk             242.329  3152     4.470  3152  648.325  3152
cant_loose          235.159  1194    10.717  1194 1481.652  1194
champions            17.142  1920     8.965  1920 1410.709  1920
hibernating         247.426  3589     2.391  3589  362.583  3589
loyal_customers      82.558  3375     8.356  3375 1216.257  3375
need_attention      113.037   806     3.739   806  553.437   806
new_customers        17.976   673     2.000   673  344.049   673
potential_loyalists  36.870  2925     3.311  2925  533.741  2925
promising            58.695   668     2.000   668  334.153   668

!["RESIM"](seg_map.png)

Yorumlamak için 4 Segment seçiyorum:
    
   ### **at_Risk:**
- Bu grupta 3152 kişi vardır.
- Ortalama olarak en son satın alma işlemini 242 gün 8 saat önce gerçekleştirmişlerdir.
- Ortalama 4.47 kez alışveriş yapmışlardır.
- Ortalama 648.32 birim para kazandırmışlardır.
- at_risk grubu satın alma davranışını uzun zaman önce yapan ve satın alma sıklığı orta olup kaybedilme riski olan sınıftır.
### **cant_loose:**
- Bu grupta 1194 kişi vardır.
- Ortalama olarak en son satın alma işlemini 235 gün 4 saat önce gerçekleştirmişlerdir.
- Ortalama 10.71 kez alışveriş yapmışlardır.
- Ortalama 1481.65 birim para kazandırmışlardır.
cant_loose grubu satın alma davranışı üzerinden uzun zaman geçen  ve satın alma sıklığı en yüksek olan sınıfdır.
### **promising:**
- Bu grupta 668 kişi vardır.
- Ortalama olarak en son satın alma işlemini 58 gün 17 saat önce gerçekleştirmişlerdir.
- Ortalama 2 kez alışveriş yapmışlardır.
- Ortalama 344.15 birim para kazandırmışlardır.
 promising grubu satın alma davranışı üzerinden uzun zaman geçmeyen ve satın alma sıklığı az fakat gelecekte iyi müşteri potansiyeli yüksek olan sınıftır.
### **new_customers:**
- Bu grupta 673 kişi vardır.
- Ortalama olarak en son satın alma işlemini 18 gün önce gerçekleştirmişlerdir.
- Ortalama 2 sıklığında alışveriş yapmışlardır.
- Ortalama 344.04 birim para kazandırmışlardır.

**RFM analizi yardımıyla yukarıda seçilen 4 segmentte yer alan müşteriler için yapılabilecek olası çalışmalar aşağıda verilmiştir:**

- Kaybedilmemesi gereken (cant_loose_them)ve riskte olan(at_Risk) müşterileri tekrar aktive edebilmeye yönelik,yapılacak kampanyada kulanılmak üzere bu müşterilerin geçmiş alışveriş profillerine uygun
olarak %20 indirim ve 1 ay içerisinde 5000 tl ve üzeri alışveriş yapmaları halinde, alışveriş tutarlarının
%35'i  oranında hediye çeki müşterilerinin kartlarına tanımlanacaktır. Kaybedilmemesi gereken ve riskli müşterilerin geçmiş alışveriş davranışlarına uygun olarak düzenlenen bu kampanya ile ilgili müşterilerin tekrar aktive olması ve firmamıza kazandırılması amaçlanmaktadır.Bu müşterilerin iletişim bilgilerine ulaşmak için müşteri_id numaraları raporlanması gerekmektedir.

In [25]:
rfm.reset_index().head()

master_id  recency  frequency  monetary  \
0  00016786-2f5a-11ea-bb80-000d3a38a36f       10      5.000   776.070   
1  00034aaa-a838-11e9-a2fc-000d3a38a36f      298      3.000   269.470   
2  000be838-85df-11ea-a90b-000d3a38a36f      213      4.000   722.690   
3  000c1fe2-a8b7-11ea-8479-000d3a38a36f       27      7.000   874.160   
4  000f5e3e-9dde-11ea-80cd-000d3a38a36f       20      7.000  1620.330   

  recency_score frequency_score monetary_score RF_SCORE      segment  
0             5               4              4       54    champions  
1             1               2              1       12  hibernating  
2             2               3              4       23      at_Risk  
3             5               4              4       54    champions  
4             5               4              5       54    champions

In [43]:
rfm_risk = rfm.loc[rfm["segment"].isin(["cant_loose","at_Risk"])].reset_index()
rfm_risk.sample(10)

master_id  recency  frequency  monetary  \
788   34dfa97c-ac38-11e9-a2fc-000d3a38a36f      288      5.000   500.450   
647   2b2213ca-9db5-11e9-9897-000d3a38a36f      353     11.000  1230.050   
397   19e8adf0-b00f-11e9-9757-000d3a38a36f      231      9.000   965.850   
2700  b624c8a8-a6db-11e9-a2fc-000d3a38a36f      308      5.000   394.930   
3440  d8ce249e-af64-11e9-a2fc-000d3a38a36f      172      8.000  1322.860   
1931  824af6d0-af4e-11e9-a2fc-000d3a38a36f      253      6.000   497.900   
2797  bbf51432-a581-11e9-a2fc-000d3a38a36f      168      6.000   929.270   
3622  dfbbece6-9825-11ea-9d91-000d3a38a36f      153      4.000   680.920   
4123  f558aac6-2b3d-11ea-9d27-000d3a38a36f      226      4.000   566.430   
803   35cc42fc-ab96-11e9-a2fc-000d3a38a36f      153      7.000  1077.990   

     recency_score frequency_score monetary_score RF_SCORE     segment  
788              1               4              3       14     at_Risk  
647              1               5              5       15  cant_loose  
397              2               5              4       25  cant_loose  
2700             1               4              2       14     at_Risk  
3440             2               5              5       25  cant_loose  
1931             1               4              3       14     at_Risk  
2797             2               4              4       24     at_Risk  
3622             2               3              4       23     at_Risk  
4123             2               4              3       24     at_Risk  
803              2               5              5       25  cant_loose

In [44]:
rfm_risk.shape[0]

4346

İstenen özelliklere uygun 4346 müşteri bulunmaktadır.

In [39]:
rfm_risk["master_id"].to_csv('risk_customer_id.csv')#master_idleri csv dosyasına kaydeder.

- Umut verici müşterilere(promising) ve yeni müşterilere(new_customers) yönelik düzenlenecek kampanya da müşterilerin firmamıza alışmaları daha sık alışveriş yapmalarını sağlamak amaçlı 1 ay içerisinde, farklı günlerde 3 kez alışveriş yapmaları halinde alışveriş tutarlarının ;en düşük olanı kadar tutar müşteri kartlarına puan olarak yüklenecektir.Bu segmentte yer alan müşterilerin iletişim bilgilerine ulaşmak için müşteri_id numaraları raporlanması gerekmektedir.

In [40]:
rfm_new = rfm.loc[rfm["segment"].isin(["promising","new_customers"])].reset_index()
rfm_new.sample(10)

master_id  recency  frequency  monetary  \
420   4307ebf0-25e5-11ea-8e1c-000d3a38a36f        7      2.000   222.970   
667   6ab8388c-b1cb-11e9-89fa-000d3a38a36f       32      2.000   171.970   
1151  bbb5c034-5bd4-11ea-a24a-000d3a38a36f        6      2.000   467.880   
714   733ff2de-4e5a-11ea-99ba-000d3a38a36f        9      2.000   392.870   
933   95b66ac2-b016-11eb-bbf8-000d3a38a36f        5      2.000   659.430   
843   88599a5a-5272-11ea-956d-000d3a38a36f       76      2.000   249.980   
564   5b2d1200-a435-11e9-a2fc-000d3a38a36f       63      2.000   323.890   
1116  b3da5dba-a52d-11e9-a2fc-000d3a38a36f       32      2.000   377.980   
624   640b9146-13d7-11ea-a001-000d3a38a36f       24      2.000   540.860   
988   9f5eadd0-26df-11ea-aaaa-000d3a38a36f       44      2.000   213.460   

     recency_score frequency_score monetary_score RF_SCORE        segment  
420              5               1              1       51  new_customers  
667              5               1              1       51  new_customers  
1151             5               1              3       51  new_customers  
714              5               1              2       51  new_customers  
933              5               1              3       51  new_customers  
843              4               1              1       41      promising  
564              4               1              2       41      promising  
1116             5               1              2       51  new_customers  
624              5               1              3       51  new_customers  
988              4               1              1       41      promising

In [41]:
rfm_new.shape[0]

1341

İstenen özelliklere uygun 1341 müşteri bulunmaktadır.

In [37]:
rfm_new["master_id"].to_csv('risk_customer_id.csv')